cc_library/src/sciler/scclib/device.py

```diff
@@ -16,14 +21,48 @@ def get_status(self):
     @abstractmethod
     def perform_instruction(self, contents):
         """
-        Defines how instructions are handled.
+        Defines how instructions are handled,
+        for all instructions defined in output of device in config.
         :param contents: contains instruction tag and calls the appropriate functions.
         :return boolean: True if instruction was valid and False if illegal instruction
         was sent or error occurred such that instruction could not be performed.
+        Returns tuple, with boolean and None if True and the failed action if false.
```

Comment:
maybe change contents to action or instruction, and specify what this dictionary can contain (instruction, component_id, value).

Reply:
sure, you do that